src/number_theory/sum_two_squares.lean

```diff
@@ -5,27 +5,227 @@ Authors: Chris Hughes
+  have H : (u * y) * (u * y) - (-1) = n * (-v ^ 2 * n + u ^ 2 + 2 * v) :=
+    by linear_combination -u ^ 2 * h + (n * v - u * x - 1) * huv,
+  refine ⟨u * y, _⟩,
+  norm_cast,
+  rw (by push_cast : (-1 : zmod n.nat_abs) = (-1 : ℤ)),
+  exact (zmod.int_coe_eq_int_coe_iff_dvd_sub _ _ _).mpr (int.nat_abs_dvd.mpr ⟨_, H⟩),
+end
+
+/-- If the natural number `n` is a sum of two squares of coprime natural numbers, then
+`-1` is a square modulo `n`. -/
+lemma zmod.is_square_neg_one_of_eq_sq_add_sq_of_coprime {n x y : ℕ} (h : n = x ^ 2 + y ^ 2)
+  (hc : x.coprime y) : is_square (-1 : zmod n) :=
+begin
+  zify at *,
+  exact zmod.is_square_neg_one_of_eq_sq_add_sq_of_is_coprime h hc.is_coprime,
+end
+
+/-- A natural number `n` is a sum of two squares if and only if `n = a^2 * b` with natural
+numbers `a` and `b` such that `-1` is a square modulo `b`. -/
+lemma nat.eq_sq_add_sq_iff_eq_sq_mul {n : ℕ} :
+  (∃ x y : ℕ, n = x ^ 2 + y ^ 2) ↔ ∃ a b : ℕ, n = a ^ 2 * b ∧ is_square (-1 : zmod b) :=
+begin
+  split,
+  { rintros ⟨x, y, h⟩,
+    by_cases hxy : x = 0 ∧ y = 0,
+    { exact ⟨0, 1, by rw [h, hxy.1, hxy.2, zero_pow zero_lt_two, add_zero, zero_mul],
+             ⟨0, by rw [zero_mul, neg_eq_zero, fin.one_eq_zero_iff]⟩⟩, },
+    { have hg := nat.pos_of_ne_zero (mt nat.gcd_eq_zero_iff.mp hxy),
+      obtain ⟨g, x₁, y₁, h₁, h₂, h₃, h₄⟩ := nat.exists_coprime' hg,
+      exact ⟨g, x₁ ^ 2 + y₁ ^ 2, by {rw [h, h₃, h₄], ring},
+             zmod.is_square_neg_one_of_eq_sq_add_sq_of_coprime rfl h₂⟩, } },
+  { rintros ⟨a, b, h₁, h₂⟩,
+    obtain ⟨x', y', h⟩ := nat.eq_sq_add_sq_of_is_square_mod_neg_one h₂,
+    exact ⟨a * x', a * y', by {rw [h₁, h], ring}⟩, }
+end
+
+end neg_one_square
+
+/-!
+### Characterization in terms of the prime factorization
+-/
+
+section main
+
+/-- A (positive) natural number `n` is a sum of two squares if and only if the exponent of
+every prime `q` such that `q%4 = 3` in the prime factorization of `n` is even.
+(The assumption `0 < n` is not present, since for `n = 0`, both sides are satisfied;
+the right hand side holds, since `padic_val_nat q 0 = 0` by definition.) -/
+lemma nat.eq_sq_add_sq_iff {n : ℕ} :
+  (∃ x y : ℕ, n = x ^ 2 + y ^ 2) ↔
+    ∀ {q : ℕ} (hq : q.prime) (h : q % 4 = 3), even (padic_val_nat q n) :=
+begin
+  rcases n.eq_zero_or_pos with rfl | hn₀,
+  { exact ⟨λ H q hq h, (@padic_val_nat.zero q).symm ▸ even_zero, λ H, ⟨0, 0, rfl⟩⟩, },
+  -- now `0 < n`
+  rw nat.eq_sq_add_sq_iff_eq_sq_mul,
+  refine ⟨λ H q hq h, _, λ H, _⟩,
+  { obtain ⟨a, b, h₁, h₂⟩ := H,
+    have hqb := padic_val_nat.eq_zero_of_not_dvd
+                  (λ hf, (hq.mod_four_ne_three_of_dvd_is_square_neg_one hf h₂) h),
+    have hab : a ^ 2 * b ≠ 0 := h₁ ▸ hn₀.ne',
+    have ha₂ := left_ne_zero_of_mul hab,
+    have ha := mt sq_eq_zero_iff.mpr ha₂,
+    have hb := right_ne_zero_of_mul hab,
+    haveI hqi : fact q.prime := ⟨hq⟩,
+    -- Using `rw` necessitates giving either `q` or `hqi` to `padic_val_nat.mul` ...
```

Comment:
I don't think it's worth highlighting that `rw` fails to pick up the `hqi` typeclass parameter (since this is not especially unusual).
```suggestion
```